File: addons/terrabrush/Scripts/EditorResources/ZonesResource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Godot;

namespace TerraBrush;

[Tool]
[GlobalClass]
public partial class ZonesResource : Resource {
    private HashSet<ImageTexture> _dirtyImageTextures = new();

    private Texture2DArray _lockTextures = new();
    private Texture2DArray _heightmapTextures = new();
    private Texture2DArray _splatmapsTextures = new();
    private Texture2DArray[] _foliagesTextures;
    private Texture2DArray _objectsTextures = new();
    private Texture2DArray _waterTextures = new();
    private Texture2DArray _snowTextures = new();
    private ImageTexture _zonesMap = new();

    public Texture2DArray LockTextures => _lockTextures;
    public Texture2DArray HeightmapTextures => _heightmapTextures;
    public Texture2DArray SplatmapsTextures => _splatmapsTextures;
    public Texture2DArray[] FoliagesTextures => _foliagesTextures;
    public Texture2DArray ObjectsTextures => _objectsTextures;
    public Texture2DArray WaterTextures => _waterTextures;
    public Texture2DArray SnowTextures => _snowTextures;
    public ImageTexture ZonesMap => _zonesMap;

    [Export] public ZoneResource[] Zones { get;set; }

    public void UpdateLockTexture(int zoneSize) {
        var images = Zones.Select(zone => zone.LockTexture?.GetImage() ?? GodotAgnostic.ImageCreateEmpty(zoneSize, zoneSize, false, Image.Format.Rf)).ToArray();

		if (images.Length != 0) {
            _lockTextures.CreateFromImages(new Godot.Collections.Array<Image>(images));
        }
    }

    public void UpdateHeightmaps() {
		var images = Zones.Select(zone => zone.HeightMapTexture.GetImage()).ToArray();
		if (images.Length != 0) {
            _heightmapTextures.CreateFromImages(new Godot.Collections.Array<Image>(images));
        }
    }

    public void UpdateSplatmapsTextures() {
        var images = Zones.Aggregate(new List<Image>(), (source, zone) => {
            if (zone.SplatmapsTexture != null) {
                source.AddRange(zone.SplatmapsTexture.Select(texture => texture.GetImage()));
            }

            return source;
        });

        if (images.Count > 0) {
            _splatmapsTextures.CreateFromImages(new Godot.Collections.Array<Image>(images));
        }
    }

    public void InitializeFoliageTextures(TerraBrush terraBrush) {
        _foliagesTextures = terraBrush.Foliages?.Select(_ => new Texture2DArray()).ToArray();
    }

    public void UpdateFoliagesTextures() {
		if (_foliagesTextures?.Length <= 0) return;

		for (var i = 0; i < _foliagesTextures?.Length; i++) {
			UpdateFoliagesTextures(i);
		}
	}

    public void UpdateFoliagesTextures(int foliageIndex) {
        var images = Zones.Select(zone => zone.FoliagesTexture[foliageIndex].GetImage()).ToArray();

        if (images.Length > 0) {
            _foliagesTextures[foliageIndex].CreateFromImages(new Godot.Collections.Array<Image>(images));
        }
    }

    public void UpdateObjectsTextures() {
        var images = Zones.Aggregate(new List<Image>(), (source, zone) => {
            if (zone.ObjectsTexture != null) {
                source.AddRange(zone.ObjectsTexture.Select(texture => texture.GetImage()));
            }

            return source;
        });

        if (images.Count > 0) {
            _objectsTextures.CreateFromImages(new Godot.Collections.Array<Image>(images));
        }
    }

    public void UpdateWaterTextures() {
        if (Zones.Any(zone => zone.WaterTexture == null)) {
            return;
        }

        _waterTextures.CreateFromImages(new Godot.Collections.Array<Image>(Zones.Select(zone => zone.WaterTexture.GetImage())));
    }

    public void UpdateZoneWaterTexture(ZoneResource zone) {
        _waterTextures.UpdateLayer(zone.WaterTexture.GetImage(), Array.IndexOf(Zones, zone));
    }

    public void UpdateSnowTextures() {
        if (Zones.Any(zone => zone.SnowTexture == null)) {
            return;
        }

        _snowTextures.CreateFromImages(new Godot.Collections.Array<Image>(Zones.Select(zone => zone.SnowTexture.GetImage())));
    }

    public void UpdateZoneSnowTexture(ZoneResource zone) {
        _snowTextures.UpdateLayer(zone.SnowTexture.GetImage(), Array.IndexOf(Zones, zone));
    }

    public void SaveResources() {
        foreach (var dirtyImageResource in _dirtyImageTextures) {
            SaveImageResource(dirtyImageResource);
        }

        _dirtyImageTextures.Clear();
    }

    private void SaveImageResource(ImageTexture image) {
        if (!string.IsNullOrWhiteSpace(image.ResourcePath) && FileAccess.FileExists(image.ResourcePath)) {
            ResourceSaver.Save(image, image.ResourcePath);
        }
    }

    public void UpdateZonesMap() {
        var zonePositions = Zones.Select(zone => zone.ZonePosition).ToArray();
		var maxX = zonePositions.Max(x => Math.Abs(x.X));
		var maxY = zonePositions.Max(x => Math.Abs(x.Y));

		var zonesMap = GodotAgnostic.ImageCreateEmpty((maxX * 2) + 1, (maxY * 2) + 1, false, Image.Format.Rf);
		zonesMap.Fill(new Color(-1, 0, 0, 0));
		for (var i = 0; i < zonePositions.Length; i++) {
			var position = zonePositions[i];
			zonesMap.SetPixel(position.X + maxX, position.Y + maxY, new Color(i, 0, 0, 0));
		}

        _zonesMap.SetImage(zonesMap);
    }

    public ZoneResource AddNewZone(TerraBrush terraBrush, Vector2I zonePosition) {
        var zone = new ZoneResource() {
            ZonePosition = zonePosition
        };

        zone.InitializeImagesForTerrain(terraBrush);

        var newList = new List<ZoneResource>(Zones) {
            zone
        };
        Zones = newList.ToArray();

        UpdateImageTextures(terraBrush.ZonesSize);

        terraBrush.Terrain?.Clipmap?.UpdateAABB();
        terraBrush.Water?.Clipmap?.UpdateAABB();
        terraBrush.Snow?.Clipmap?.UpdateAABB();

        return zone;
    }

    public void AddDirtyImageTexture(ImageTexture imageTexture) {
        _dirtyImageTextures.Add(imageTexture);
    }

    public void UpdateImageTextures(int zoneSize) {
        if (Engine.IsEditorHint()) {
            UpdateLockTexture(zoneSize);
        }
        UpdateHeightmaps();
        UpdateSplatmapsTextures();
        UpdateFoliagesTextures();
        UpdateObjectsTextures();
        UpdateWaterTextures();
        UpdateSnowTextures();
        UpdateZonesMap();
    }

    public ZoneResource GetZoneForZoneInfo(ZoneInfo zoneInfo) {
        return Zones?.FirstOrDefault(x => x.ZonePosition.X == zoneInfo.ZonePosition.X && x.ZonePosition.Y == zoneInfo.ZonePosition.Y);
    }
}

File: src/editor_resources/zones_resource.cpp
#include "zones_resource.h"
#include "../misc/zone_info.h"

#include <godot_cpp/classes/image.hpp>
#include <godot_cpp/classes/file_access.hpp>
#include <godot_cpp/classes/resource_saver.hpp>
#include <godot_cpp/classes/engine.hpp>
#include <godot_cpp/classes/texture2d_array.hpp>

using namespace godot;

void ZonesResource::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_zones"), &ZonesResource::get_zones);
    ClassDB::bind_method(D_METHOD("set_zones", "value"), &ZonesResource::set_zones);
    ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "zones", PROPERTY_HINT_TYPE_STRING, "24/17:ZoneResource"), "set_zones", "get_zones");
}

ZonesResource::ZonesResource() {
    _dirtyImages = std::unordered_set<Ref<Image>>();

    _lockTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _heightmapTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _colorTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _splatmapsTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _foliagesTextures = TypedArray<Ref<Texture2DArray>>();
    _objectsTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _waterTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _snowTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _metaInfoTextures = Ref<Texture2DArray>(memnew(Texture2DArray));
    _zonesMap = Ref<ImageTexture>(memnew(ImageTexture));

    _zones = TypedArray<Ref<ZoneResource>>();
}

ZonesResource::~ZonesResource() {}

Ref<Texture2DArray> ZonesResource::get_lockTextures() const {
    return _lockTextures;
}
Ref<Texture2DArray> ZonesResource::get_heightmapTextures() const {
    return _heightmapTextures;
}
Ref<Texture2DArray> ZonesResource::get_colorTextures() const {
    return _colorTextures;
}
Ref<Texture2DArray> ZonesResource::get_splatmapsTextures() const {
    return _splatmapsTextures;
}
TypedArray<Ref<Texture2DArray>> ZonesResource::get_foliagesTextures() const {
    return _foliagesTextures;
}
Ref<Texture2DArray> ZonesResource::get_objectsTextures() const {
    return _objectsTextures;
}
Ref<Texture2DArray> ZonesResource::get_waterTextures() const {
    return _waterTextures;
}
Ref<Texture2DArray> ZonesResource::get_snowTextures() const {
    return _snowTextures;
}
Ref<Texture2DArray> ZonesResource::get_metaInfoTextures() const {
    return _metaInfoTextures;
}
Ref<ImageTexture> ZonesResource::get_zonesMap() const {
    return _zonesMap;
}

TypedArray<Ref<ZoneResource>> ZonesResource::get_zones() const {
    return _zones;
}
void ZonesResource::set_zones(const TypedArray<Ref<ZoneResource>> value) {
    _zones = value;
}

void ZonesResource::updateLockTexture(int zoneSize) {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        if (zone->get_lockTexture().is_null()) {
            images.append(Image::create_empty(zoneSize, zoneSize, false, Image::Format::FORMAT_RF));
        } else {
            images.append(zone->get_lockTexture());
        }
    }

    if (images.size() > 0) {
        _lockTextures->create_from_images(images);
    }
}

void ZonesResource::updateHeightmaps() {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        images.append(zone->get_heightMapImage());
    }

    if (images.size() > 0) {
        _heightmapTextures->create_from_images(images);
    }
}

void ZonesResource::updateColorTextures(int zoneSize) {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        if (zone->get_colorImage().is_null()) {
            images.append(Image::create_empty(zoneSize, zoneSize, false, Image::Format::FORMAT_RGBA8));
        } else {
            images.append(zone->get_colorImage());
        }
    }

    if (images.size() > 0) {
        _colorTextures->create_from_images(images);
    }
}

void ZonesResource::updateSplatmapsTextures() {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        if (zone->get_splatmapsImage().size() > 0) {
            for (Ref<Image> splatmap : zone->get_splatmapsImage()) {
                images.append(splatmap);
            }
        }
    }

    if (images.size() > 0) {
        _splatmapsTextures->create_from_images(images);
    }
}

void ZonesResource::initializeFoliageTextures(int foliageCount) {
    _foliagesTextures = TypedArray<Ref<Texture2DArray>>();
    for (int i = 0; i < foliageCount; i++) {
        _foliagesTextures.append(memnew(Texture2DArray));
    }
}

void ZonesResource::updateFoliagesTextures() {
    if (_foliagesTextures.size() <= 0) return;

    for (int i = 0; i < _foliagesTextures.size(); i++) {
        updateFoliagesTextures(i);
    }
}

void ZonesResource::updateFoliagesTextures(int foliageIndex) {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        TypedArray<Ref<Image>> foliagesTexture = zone->get_foliagesImage();
        images.append(Ref<Image>(foliagesTexture[foliageIndex]));
    }

    if (images.size() > 0) {
        Ref<Texture2DArray>(_foliagesTextures[foliageIndex])->create_from_images(images);
    }
}

void ZonesResource::updateObjectsTextures() {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        if (zone->get_objectsImage().size() > 0) {
            for (Ref<Image> objectTexture : zone->get_objectsImage()) {
                images.append(objectTexture);
            }
        }
    }

    if (images.size() > 0) {
        _objectsTextures->create_from_images(images);
    }
}

void ZonesResource::updateWaterTextures() {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        if (!zone->get_waterImage().is_null()) {
            images.append(zone->get_waterImage());
        }
    }

    if (images.size() > 0) {
        _waterTextures->create_from_images(images);
    }
}

void ZonesResource::updateZoneWaterTexture(Ref<ZoneResource> zone) {
    int zoneIndex = _zones.find(zone);
    _waterTextures->update_layer(zone->get_waterImage(), zoneIndex);
}

void ZonesResource::updateSnowTextures() {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        if (!zone->get_snowImage().is_null()) {
            images.append(zone->get_snowImage());
        }
    }

    if (images.size() > 0) {
        _snowTextures->create_from_images(images);
    }
}

void ZonesResource::updateZoneSnowTexture(Ref<ZoneResource> zone) {
    int zoneIndex = _zones.find(zone);
    _snowTextures->update_layer(zone->get_snowImage(), zoneIndex);
}

void ZonesResource::updateMetaInfoTextures() {
    TypedArray<Ref<Image>> images = TypedArray<Ref<Image>>();
    for (Ref<ZoneResource> zone : _zones) {
        if (!zone->get_metaInfoImage().is_null()) {
            images.append(zone->get_metaInfoImage());
        }
    }

    if (images.size() > 0) {
        _metaInfoTextures->create_from_images(images);
    }
}

void ZonesResource::updateZoneMetaInfoTexture(Ref<ZoneResource> zone) {
    int zoneIndex = _zones.find(zone);
    _metaInfoTextures->update_layer(zone->get_metaInfoImage(), zoneIndex);
}

void ZonesResource::saveResources() {
    if (_dirtyImages.size() == 0) {
        return;
    }

    for (Ref<Image> dirtyImageResource : _dirtyImages) {
        saveImageResource(dirtyImageResource);
    }

    _dirtyImages.clear();
}

void ZonesResource::saveImageResource(Ref<Image> image) {
    if (!image->get_path().is_empty() && FileAccess::file_exists(image->get_path())) {
        ResourceSaver::get_singleton()->save(image, image->get_path());
    }
}

void ZonesResource::updateZonesMap() {
    TypedArray<Vector2> zonePositions = TypedArray<Vector2>();
    int maxX = 0;
    int maxY = 0;
    for (Ref<ZoneResource> zone : _zones) {
        Vector2 zonePosition = zone->get_zonePosition();
        zonePositions.append(zonePosition);

        maxX = Math::max(maxX, (int) Math::abs(zonePosition.x));
        maxY = Math::max(maxY, (int) Math::abs(zonePosition.y));
    }

    Ref<Image> zonesMap = Image::create_empty((maxX * 2) + 1, (maxY * 2) + 1, false, Image::Format::FORMAT_RF);
    zonesMap->fill(Color(-1, 0, 0, 0));
    for (int i = 0; i < zonePositions.size(); i++) {
        Vector2 position = zonePositions[i];
        zonesMap->set_pixel(position.x + maxX, position.y + maxY, Color(i, 0, 0, 0));
    }

    _zonesMap->set_image(zonesMap);
}

void ZonesResource::addDirtyImage(Ref<Image> imageTexture) {
    _dirtyImages.insert(imageTexture);
}

void ZonesResource::updateImageImages(int zoneSize) {
    if (Engine::get_singleton()->is_editor_hint()) {
        updateLockTexture(zoneSize);
    }
    updateHeightmaps();
    updateColorTextures(zoneSize);
    updateSplatmapsTextures();
    updateFoliagesTextures();
    updateObjectsTextures();
    updateWaterTextures();
    updateSnowTextures();
    updateMetaInfoTextures();
    updateZonesMap();
}

Ref<ZoneResource> ZonesResource::getZoneForZoneInfo(ZoneInfo zoneInfo) {
    if (_zones.size() == 0) {
        return nullptr;
    }

    for (Ref<ZoneResource> zone : _zones) {
        if (zone.is_null()) {
            return nullptr;
        }

        if (zone->get_zonePosition().x == zoneInfo.zonePosition.x && zone->get_zonePosition().y == zoneInfo.zonePosition.y) {
            return zone;
        }
    }

    return nullptr;
}